disjoint_set/main.py

```diff
@@ -54,9 +61,15 @@ def find(self, x: T) -> T:
         >>> ds.find(1)
         2
         """
-        if x != self._data[x]:
-            self._data[x] = self.find(self._data[x])
-        return self._data[x]
+        # EACH node along the path should be pointed to the ROOT, instead of its DIRECT PARENT.
+        # see unittest case for detail
+        path = []
+        while x != self._data[x]:  # loop till the ROOT
+            path.append(x)
+            x = self._data[x]
+        for e in path:  # pointing EACH NODE directly to ROOT
+            self._data[e] = x
+        return x
```

Comment:
I guess we don't even need to keep track of list. 
```suggest 
        while x != self._data[x]:
            self._data[x] = self._data[self._data[x]]
            x = self._data[x]
        return x
```
I didn't test it thoroughly but seems to work.

Reply:
Thanks for the suggestion. I ended up using the above solution in the latest package version!